docs/guides/ajax.rst

```diff
@@ -1,31 +1,220 @@
+#. The request is made to the server, either rendering a view or a form, calling an action, or loading a path.
+#. Echoed JSON is turned into a response object (``Elgg\Services\AjaxResponse``).
+#. The response object is filtered by the hook ``ajax_response``.
+#. The response object is used to create the HTTP response.
+#. Client-side, the response data is filtered by the hook ``ajax_response_data``.
+#. The method returns a ``jqXHR`` object, which can be used as a Promise.
 
-    $arg1 = (int)get_input('arg1');
-    $arg2 = (int)get_input('arg2');
+More notes:
 
-    system_message('We did it!');
+* By default the ``elgg/spinner`` module is automatically used during requests.
+* User messages generated by ``system_message()`` and ``register_error()`` are collected and displayed on the client.
+* Elgg gives you a default error handler that shows a generic message if output fails.
+* PHP exceptions or denied resource return HTTP error codes, resulting in use of the client-side error handler.
+* The default HTTP method is POST for actions, otherwise GET. You can set it via ``options.method``.
+* For client caching, set ``options.method`` to "GET" and ``options.data.response_ttl`` to the max-age you want in seconds.
+* Do not use ``forward()`` with this API.
 
-    echo json_encode([
-        'sum' => $arg1 + $arg2,
-        'product' => $arg1 * $arg2,
-    ]);
+Performing actions
+------------------
+
+To execute an action, use ``performAction('action_name', options)``:
+
+.. code-block:: php
+
+	// in myplugin/actions/do_math.php
+
+	elgg_ajax_gatekeeper();
+
+	$arg1 = (int)get_input('arg1');
+	$arg2 = (int)get_input('arg2');
+
+	// will be rendered client-side
+	system_message('We did it!');
+
+	echo json_encode([
+		'sum' => $arg1 + $arg2,
+		'product' => $arg1 * $arg2,
+	]);
+
+.. code-block:: js
+
+	var ajax = new require('elgg/Ajax')();
+
+	ajax.performAction('do_math', {
+		data: {
+			arg1: 1,
+			arg2: 2
+		},
+	}).done(function (output) {
+		alert(output.sum);
+		alert(output.product);
+	});
+
+Notes for actions:
+
+* All hooks have type ``action:<action_name>``. In this case, "action:do_math".
+* CSRF tokens are added to the data.
+* The default method is POST.
+
+
+Fetching data
+-------------
+
+To fetch arbitrary data from registered page handlers, use ``fetchPath('url_path', options)``.
+
+.. code-block:: php
+
+	// in myplugin/start.php
+	elgg_register_page_handler('myplugin_time', 'myplugin_get_time');
+
+	function myplugin_get_time() {
+		elgg_ajax_gatekeeper();
+
+		echo json_encode([
+			'rfc2822' => date(DATE_RFC2822),
+			'day' => date('l'),
+		]);
+
+		return true;
+	}
+
+.. code-block:: js
+
+	var ajax = new require('elgg/Ajax')();
+
+	ajax.fetchPath('myplugin_time').done(function (output) {
+		alert(output.rfc2822);
+		alert(output.day);
+	});
+
+Notes for paths:
+
+* All hooks have type ``path:<url_path>``. In this case, "path:<myplugin_time>".
```

Comment:
I definitely find it more useful than `fetchView`, as I avoid ajax views as much as can. I believe that anything loaded via ajax should have a fully qualifying URL and should be reachable by the browser should JS fail to hijack a click action.
Perhaps we should allow fetching a full path, but only trigger a hook on the page handler ID?